csharp/Program.cs

```diff
@@ -4,38 +4,34 @@
 using System.Text.Json;
 using System.Text.Json.Serialization;
 
+
 const int topN = 5;
-var posts = JsonSerializer.Deserialize<List<Post>>(File.ReadAllText(@"../posts.json"), MyJsonContext.Default.ListPost);
+var posts = JsonSerializer.Deserialize<List<Post>>(File.ReadAllText(@"../posts.json"), MyJsonContext.Default.ListPost)!;
+var postsCount = posts.Count;
 
 var sw = Stopwatch.StartNew();
 
-// slower when int[] is used
-var tagMapTemp = new Dictionary<string, Stack<int>>(100);
+var tagMap = new Dictionary<string, int[]>(100);
+
 
-for (var i = 0; i < posts!.Count; i++)
+for (var i = 0; i < postsCount; i++)
 {
     foreach (var tag in posts[i].Tags)
     {
         // single lookup
-        ref var stack = ref CollectionsMarshal.GetValueRefOrAddDefault(tagMapTemp, tag, out _);
-        stack ??= new Stack<int>();
-        stack.Push(i);
+        ref var stack = ref CollectionsMarshal.GetValueRefOrAddDefault(tagMap, tag, out _);
+        stack ??= new int[100];
+        _ = stack.Append(i);
```

Comment:
This looks like a bug. Append is extension method that creates a new IEnumerable instance and you are throwing the result away while the source array stays unmodified.

Reply:
That version is outdated

Reply:
Yes, was a bug - ended up reverting back to the Stack that was being used - swapped it out for a LinkedList instead of a stack though in the end cause, LL is very, very slightly faster

Reply:
I tried LinkedList and it's slightly slower on my machine. List and Stack seem to be the same. The most impact has the conversion to array at the end (tagMapTemp->tagMap).

Reply:
Interesting, LinkedList should add things faster that a Stack will, but access / traversal is generally slower. For me it was consistently about 4ms faster on my machine. using LinkedList, but could easily be possible on different hardware etc it swings the other way

Reply:
LinkedList.AddLast is constant time, but has a larger overhead for each call. List is usually faster when adding items at the end (same for Stack.Push and Queue.Enqueue).